File: LeetCode/230.Kth_Smallest_Element_in_a_BST/Mario/230.Kth_Smallest_Element_in_a_BST.py
#!/usr/bin/env python3

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        vals = []

        def dfs(node):
            if node:
                if len(vals) == k:
                    return
                dfs(node.left)
                vals.append(node.val)
                dfs(node.right)

        dfs(root)
        return vals[k-1]

File: LeetCode/230.Kth_Smallest_Element_in_a_BST/Mario/230.Kth_Smallest_Element_in_a_BST.java
class Solution {
    // BST inorder traversal
    private List<Integer> list;

    public int kthSmallest(TreeNode root, int k) {
        list = new ArrayList<>();
        inorderTraversal(root, k);
        return list.get(k - 1);
    }

    private void inorderTraversal(TreeNode node, int k) {
        if (node == null || list.size() >= k)
            return;
        inorderTraversal(node.left, k);
        list.add(node.val);
        inorderTraversal(node.right, k);
    }
}